File: src-tauri/src/notifications/error_detector.rs
//! `error_detector` — orchestrates trigger matching across a session's parsed
//! messages. Ported from `internal/notifications/error_detector.go` (W14).

use std::collections::HashMap;

use crate::config::state::types::NotificationTrigger;
use crate::types::messages::ParsedMessage;

use super::checks::{
    check_token_threshold_trigger, check_tool_result_trigger, check_tool_use_trigger,
};
use super::tool_maps::{build_tool_result_map, build_tool_use_map, ToolResultInfo, ToolUseInfo};
use super::types::DetectedError;

/// Whether a project ID is covered by the trigger's `repositoryIds` scope.
/// Nil/empty scope means "all repositories".
fn matches_repository_scope(project_id: &str, repository_ids: Option<&[String]>) -> bool {
    match repository_ids {
        Some(ids) if !ids.is_empty() => ids.iter().any(|id| id == project_id),
        _ => true,
    }
}

/// Routes one trigger against one message to the appropriate checker.
#[allow(clippy::too_many_arguments)]
fn check_trigger(
    msg: &ParsedMessage,
    trigger: &NotificationTrigger,
    tool_use_map: &HashMap<String, ToolUseInfo>,
    tool_result_map: &HashMap<String, ToolResultInfo>,
    session_id: &str,
    project_id: &str,
    file_path: &str,
    line_number: u32,
) -> Vec<DetectedError> {
    if !matches_repository_scope(project_id, trigger.repository_ids.as_deref()) {
        return Vec::new();
    }

    if trigger.mode == "token_threshold" {
        return check_token_threshold_trigger(
            msg,
            trigger,
            tool_result_map,
            session_id,
            project_id,
            file_path,
            line_number,
        );
    }

    if trigger.content_type == "tool_result" {
        return check_tool_result_trigger(
            msg,
            trigger,
            tool_use_map,
            session_id,
            project_id,
            file_path,
            line_number,
        )
        .into_iter()
        .collect();
    }

    if trigger.content_type == "tool_use" {
        return check_tool_use_trigger(msg, trigger, session_id, project_id, file_path, line_number)
            .into_iter()
            .collect();
    }

    Vec::new()
}

/// Runs all triggers against every message in the session.
pub fn detect_errors(
    messages: &[ParsedMessage],
    session_id: &str,
    project_id: &str,
    file_path: &str,
    triggers: &[NotificationTrigger],
) -> Vec<DetectedError> {
    if triggers.is_empty() {
        return Vec::new();
    }

    let tool_use_map = build_tool_use_map(messages);
    let tool_result_map = build_tool_result_map(messages);
    let mut errors = Vec::new();

    for (i, msg) in messages.iter().enumerate() {
        let line_number = (i + 1) as u32;
        for trigger in triggers {
            errors.extend(check_trigger(
                msg,
                trigger,
                &tool_use_map,
                &tool_result_map,
                session_id,
                project_id,
                file_path,
                line_number,
            ));
        }
    }

    errors
}

/// Runs a single trigger against a session's messages.
pub fn detect_errors_with_trigger(
    messages: &[ParsedMessage],
    trigger: &NotificationTrigger,
    session_id: &str,
    project_id: &str,
    file_path: &str,
) -> Vec<DetectedError> {
    detect_errors(
        messages,
        session_id,
        project_id,
        file_path,
        std::slice::from_ref(trigger),
    )
}

#[cfg(test)]
#[path = "error_detector_tests.rs"]
mod tests;

File: internal/notifications/error_detector.go
// error_detector.go.
// Orchestrates trigger matching across a session's parsed messages.
package notifications

import (
	"claude-devtools/internal/config"
	"claude-devtools/internal/domain"
)

// matchesRepositoryScope checks whether a project ID is covered by the trigger's
// repositoryIds scope. Nil/empty scope means "all repositories".
// Mirrors error_detector.rs::matches_repository_scope.
func matchesRepositoryScope(projectID string, repositoryIDs *[]string) bool {
	if repositoryIDs == nil || len(*repositoryIDs) == 0 {
		return true
	}
	for _, id := range *repositoryIDs {
		if id == projectID {
			return true
		}
	}
	return false
}

// checkTrigger routes one trigger against one message to the appropriate checker.
// Mirrors error_detector.rs::check_trigger.
func checkTrigger(
	msg *domain.ParsedMessage,
	trigger *config.NotificationTrigger,
	toolUseMap map[string]ToolUseInfo,
	toolResultMap map[string]ToolResultInfo,
	sessionID, projectID, filePath string,
	lineNumber uint32,
) []DetectedError {
	if !matchesRepositoryScope(projectID, trigger.RepositoryIDs) {
		return nil
	}

	if trigger.Mode == "token_threshold" {
		return CheckTokenThresholdTrigger(msg, trigger, toolResultMap, sessionID, projectID, filePath, lineNumber)
	}

	if trigger.ContentType == "tool_result" {
		if e := CheckToolResultTrigger(msg, trigger, toolUseMap, sessionID, projectID, filePath, lineNumber); e != nil {
			return []DetectedError{*e}
		}
		return nil
	}

	if trigger.ContentType == "tool_use" {
		if e := CheckToolUseTrigger(msg, trigger, sessionID, projectID, filePath, lineNumber); e != nil {
			return []DetectedError{*e}
		}
		return nil
	}

	return nil
}

// DetectErrors runs all triggers against every message in the session.
// Mirrors error_detector.rs::detect_errors.
func DetectErrors(
	messages []domain.ParsedMessage,
	sessionID, projectID, filePath string,
	triggers []config.NotificationTrigger,
) []DetectedError {
	if len(triggers) == 0 {
		return []DetectedError{}
	}

	toolUseMap := BuildToolUseMap(messages)
	toolResultMap := BuildToolResultMap(messages)
	var errors []DetectedError

	for i := range messages {
		lineNumber := uint32(i + 1)
		for j := range triggers {
			errs := checkTrigger(&messages[i], &triggers[j], toolUseMap, toolResultMap, sessionID, projectID, filePath, lineNumber)
			errors = append(errors, errs...)
		}
	}

	if errors == nil {
		return []DetectedError{}
	}
	return errors
}

// DetectErrorsWithTrigger runs a single trigger against a session's messages.
// Mirrors error_detector.rs::detect_errors_with_trigger.
func DetectErrorsWithTrigger(
	messages []domain.ParsedMessage,
	trigger *config.NotificationTrigger,
	sessionID, projectID, filePath string,
) []DetectedError {
	return DetectErrors(messages, sessionID, projectID, filePath, []config.NotificationTrigger{*trigger})
}
